File: models/mediafiletype.cpp
#include "mediafiletype.h"
#include <algorithm>

using namespace NickvisionTubeConverter::Models;

MediaFileType::MediaFileType(Value fileType) : m_value{fileType}
{

}

std::optional<MediaFileType> MediaFileType::parse(const std::string& s)
{
    std::string sToLower{s};
    std::transform(sToLower.begin(), sToLower.end(), sToLower.begin(), ::tolower);
    if(sToLower.find(".mp4") != std::string::npos)
    {
        return MediaFileType::MP4;
    }
    else if(sToLower.find(".webm") != std::string::npos)
    {
        return MediaFileType::WEBM;
    }
    else if(sToLower.find(".mp3") != std::string::npos)
    {
        return MediaFileType::MP3;
    }
    else if(sToLower.find(".ogg") != std::string::npos)
    {
        return MediaFileType::OGG;
    }
    else if(sToLower.find(".flac") != std::string::npos)
    {
        return MediaFileType::FLAC;
    }
    else if(sToLower.find(".wma") != std::string::npos)
    {
        return MediaFileType::WMA;
    }
    else if(sToLower.find(".wav") != std::string::npos)
    {
        return MediaFileType::WAV;
    }
    else
    {
        return std::nullopt;
    }
}

MediaFileType::operator Value() const
{
    return m_value;
}

bool MediaFileType::operator==(const MediaFileType& toCompare) const
{
    return m_value == toCompare.m_value;
}

bool MediaFileType::operator!=(const MediaFileType& toCompare) const
{
    return m_value != toCompare.m_value;
}

bool MediaFileType::operator==(MediaFileType::Value toCompare) const
{
    return m_value == toCompare;
}

bool MediaFileType::operator!=(MediaFileType::Value toCompare) const
{
    return m_value != toCompare;
}

std::string MediaFileType::toString() const
{
    switch(m_value)
    {
        case MediaFileType::MP4:
            return "mp4";
        case MediaFileType::WEBM:
            return "webm";
        case MediaFileType::MP3:
            return "mp3";
        case MediaFileType::OGG:
            return "ogg";
        case MediaFileType::FLAC:
            return "flac";
        case MediaFileType::WMA:
            return "wma";
        case MediaFileType::WAV:
            return "wav";
        default:
            return "";
    }
}

std::string MediaFileType::toDotExtension() const
{
    switch(m_value)
    {
        case MediaFileType::MP4:
            return ".mp4";
        case MediaFileType::WEBM:
            return ".webm";
        case MediaFileType::MP3:
            return ".mp3";
        case MediaFileType::OGG:
            return ".ogg";
        case MediaFileType::FLAC:
            return ".flac";
        case MediaFileType::WMA:
            return ".wma";
        case MediaFileType::WAV:
            return ".wav";
        default:
            return "";
    }
}

bool MediaFileType::isAudio() const
{
    switch(m_value)
    {
        case MediaFileType::MP4:
            return false;
        case MediaFileType::WEBM:
            return false;
        case MediaFileType::MP3:
            return true;
        case MediaFileType::OGG:
            return true;
        case MediaFileType::FLAC:
            return true;
        case MediaFileType::WMA:
            return true;
        case MediaFileType::WAV:
            return true;
        default:
            return false;
    }
}

bool MediaFileType::isVideo() const
{
    switch(m_value)
    {
        case MediaFileType::MP4:
            return true;
        case MediaFileType::WEBM:
            return true;
        case MediaFileType::MP3:
            return false;
        case MediaFileType::OGG:
            return false;
        case MediaFileType::FLAC:
            return false;
        case MediaFileType::WMA:
            return false;
        case MediaFileType::WAV:
            return false;
        default:
            return false;
    }
}

File: NickvisionTubeConverter.Shared/Models/MediaFileType.cs
using System;

namespace NickvisionTubeConverter.Shared.Models;

/// <summary>
/// A model for media file types
/// </summary>
public enum MediaFileType
{
    MP4 = 0,
    WEBM,
    MP3,
    M4A,
    OPUS,
    FLAC,
    WAV,
    Video,
    Audio
}

/// <summary>
/// Helper functions for media file types
/// </summary>
public static class MediaFileTypeHelpers
{
    /// <summary>
    /// Parse a MediaFileType from a string
    /// </summary>
    /// <param name="s">The string to parse</param>
    /// <returns>The MediaFileType or null if failed</returns>
    public static MediaFileType? Parse(string s)
    {
        if (s.IndexOf('.') != -1)
        {
            s = s.Remove(s.IndexOf('.'), 1);
        }
        try
        {
            return (MediaFileType?)Enum.Parse(typeof(MediaFileType), s, true);
        }
        catch
        {
            return null;
        }
    }

    /// <summary>
    /// Gets the dot extension of the MediaFileType
    /// </summary>
    /// <param name="type">The MediaFileType</param>
    /// <returns>The dot extension of the MediaFileType</returns>
    public static string GetDotExtension(this MediaFileType type) => $".{type.ToString().ToLower()}";

    /// <summary>
    /// Gets whether or not the MediaFileType is an audio file type
    /// </summary>
    /// <param name="type">The MediaFileType</param>
    /// <returns>True if audio file type, else false</returns>
    public static bool GetIsAudio(this MediaFileType type) => type switch
    {
        MediaFileType.MP4 => false,
        MediaFileType.WEBM => false,
        MediaFileType.MP3 => true,
        MediaFileType.M4A => true,
        MediaFileType.OPUS => true,
        MediaFileType.FLAC => true,
        MediaFileType.WAV => true,
        MediaFileType.Video => false,
        MediaFileType.Audio => true,
        _ => false
    };

    /// <summary>
    /// Gets whether or not the MediaFileType is a media file type
    /// </summary>
    /// <param name="type">The MediaFileType</param>
    /// <returns>True if media file type, else false</returns>
    public static bool GetIsVideo(this MediaFileType type) => type switch
    {
        MediaFileType.MP4 => true,
        MediaFileType.WEBM => true,
        MediaFileType.MP3 => false,
        MediaFileType.M4A => false,
        MediaFileType.OPUS => false,
        MediaFileType.FLAC => false,
        MediaFileType.WAV => false,
        MediaFileType.Video => true,
        MediaFileType.Audio => false,
        _ => false
    };

    /// <summary>
    /// Gets whether or not the MediaFileType supports embedding thumbnails
    /// </summary>
    /// <param name="type">The MediaFileType</param>
    /// <returns>True if supported, else false</returns>
    public static bool GetSupportsThumbnails(this MediaFileType type) => type switch
    {
        MediaFileType.MP4 => true,
        MediaFileType.WEBM => false,
        MediaFileType.MP3 => true,
        MediaFileType.M4A => true,
        MediaFileType.OPUS => true,
        MediaFileType.FLAC => true,
        MediaFileType.WAV => false,
        MediaFileType.Video => false,
        MediaFileType.Audio => false,
        _ => false
    };

    /// <summary>
    /// Gets whether or not the MediaFileType is generic
    /// </summary>
    /// <param name="type">The MediaFileType</param>
    /// <returns>True if generic, else false</returns>
    public static bool GetIsGeneric(this MediaFileType type) => type switch
    {
        MediaFileType.Video => true,
        MediaFileType.Audio => true,
        _ => false
    };
}
